shared/src/event/mod.rs

```diff
@@ -0,0 +1,70 @@
+use bitflags::bitflags;
+use fvm_ipld_encoding::{Cbor, RawBytes};
+use serde::{Deserialize, Serialize};
+use serde_tuple::*;
+
+use crate::ActorID;
+
+/// Event with extra information stamped by the FVM. This is the structure that gets committed
+/// on-chain via the receipt.
+#[derive(Serialize_tuple, Deserialize_tuple, PartialEq, Eq, Clone, Debug)]
+pub struct StampedEvent {
+    /// Carries the ID of the actor that emitted this event.
+    emitter: ActorID,
+    /// The event as emitted by the actor.
+    event: ActorEvent,
+}
+
+impl Cbor for StampedEvent {}
+
+impl StampedEvent {
+    pub fn new(emitter: ActorID, event: ActorEvent) -> Self {
+        Self { emitter, event }
+    }
+}
+
+/// An event as originally emitted by the actor.
+#[derive(Serialize_tuple, Deserialize_tuple, PartialEq, Eq, Clone, Debug)]
+#[serde(transparent)]
+pub struct ActorEvent {
+    pub entries: Vec<Entry>,
+}
+
+impl Cbor for ActorEvent {}
+
+impl From<Vec<Entry>> for ActorEvent {
+    fn from(entries: Vec<Entry>) -> Self {
+        Self { entries }
+    }
+}
+
+bitflags! {
+    /// Flags associated with an Event entry.
+    #[derive(Deserialize, Serialize)]
+    #[serde(transparent)]
+    pub struct Flags: u8 {
+        const FLAG_INDEXED_KEY      = 0b00000001;
+        const FLAG_INDEXED_VALUE    = 0b00000010;
+        const FLAG_INDEXED_ALL      = Self::FLAG_INDEXED_KEY.bits | Self::FLAG_INDEXED_VALUE.bits;
+    }
+}
+
+/// A key value entry inside an Event.
+#[derive(Serialize_tuple, Deserialize_tuple, PartialEq, Eq, Clone, Debug)]
+pub struct Entry {
+    /// A bitmap conveying metadata or hints about this entry.
+    pub flags: Flags,
+    /// The key of this event.
+    pub key: String,
+    /// Any DAG-CBOR encodeable type.
+    pub value: RawBytes,
```

Comment:
It would be nice if this were an `IgnoreAny` (or `Ipld` if we wanted to actually keep it). By using `RawBytes`, we've "hidden" any internal CIDs from the codec.

But I don't feel _too_ strongly about that.

Reply:
Ok, I guess I feel somewhat strongly about this. I'd say that the value should either by "arbitrary bytes", or an _actual_ CBOR value. Having cbor as a byte-string within another cbor-value is icky.

Reply:
I agree with the concern. But in practice I'm not convinced it changes much. What's important is that the validation logic is in place somewhere. This being represented as an `Ipld::Ipld` here makes things harder for actor code (because `Entry` is a shared type). An `IgnoredAny` would simply not work for the same reason. We would need to bifurcate this type into the FVM-side (which could use `IgnoredAny` or  `Ipld::Ipld`) and the SDK-side type. And I'm not sure what the right field type would be for the latter if not `RawBytes`?

Reply:
So, these behave _completely_ differently.

- `Ipld` and `IgnoreAny` means the data would actually be cbor.
- `RawBytes` means that the data would be a byte array (that _may_ be cbor if we validate it).

The important thing is that any links in that raw data wouldn't be resolvable. I.e., any IPLD tooling would treat it as "just bytes".

Reply:
> This being represented as an Ipld::Ipld here makes things harder for actor code (because Entry is a shared type). An IgnoredAny would simply not work for the same reason.

The "right" solution is likely to make the `Event` generic over this type. Then we can use whatever type we want for the value, depending on the context.

- When validating, we can use `IgnoredAny` (once we fix it to actually validate fully, but we don't really care about validation right now anyways).
- The most "general" version is `Ipld`, it just gets expensive because it allocates a lot.
- We should implement a `RawValue` equivalent for IPLD: https://docs.rs/serde_json/latest/serde_json/value/struct.RawValue.html. We just don't have that right now.

Reply:
> Ipld and IgnoreAny means the data would actually be cbor.

It just changes the timing at which you verify. Using `Ipld` or `IgnoreAny` would give us that guarantee during deserialization. Validating explicitly gives us that guarantee later. But in all cases you would walk away guaranteeing that the input is DAG-CBOR.

> The important thing is that any links in that raw data wouldn't be resolvable. I.e., any IPLD tooling would treat it as "just bytes".

How much of a problem is that?

Inside the FVM, we never expect to resolve these links (for all we know, they could be pointing at some NFT stored in the public IPFS network). I'd argue that we have no interest in processing Entry values at all, beyond validating that they're correct DAG-CBOR. And the reason we enforce DAG-CBOR is to give observers (which will be vary varied) _some_ guarantees about the structure of the data. Otherwise, they will have zero ability to reason about the events they observe.

Reply:
> How much of a problem is that?

It breaks tooling:

- If the user linked to data, pinning a receipt wouldn't pin any data linked-to by an event.
- Selectors wouldn't be able to see inside events.

> I'd argue that we have no interest in processing Entry values at all, beyond validating that they're correct DAG-CBOR.

Either we care or we don't.

- If we care, it should be an embedded CBOR object. E.g., if we ever want to maybe consume events on-chain.
- If we don't, we can say that the user _should_ use CBOR, but I see no reason to spend extra time validating (because it doesn't actually _matter_ to us).

Given this discussion, I'm leaning towards the second.

Reply:
> Given this discussion, I'm leaning towards the second.

Which is basically what we have, just with an extra "we're never going to validate this" guarantee.

Reply:
Hm, how about we table this discussion until I get around to updating the FIP? Then we can discuss with a larger audience there.

In short, I still think it's beneficial to enforce DAG-CBOR as these events are meant for external observability, and not having a consistent data format hinders such observability. I do not think we should **reason** _at all_ about these payloads. No link discovery, no reachability analysis, nothing. Validating structural correctness is where I'd draw the line in terms of FVM responsibility.

If we don't reason about these, I'd argue it's not a requirement to use types that are transparent to IPLD tooling (e.g. `Ipld::Ipld`, I don't know about `IgnoredAny`) in the implementation. We may opt to do so for technical reasons, though, not arguing that. But bearing in mind that that would require more type sophistication here (I can sense your eyes light up 😆).